File: Typescript/nn.ts
import * as tf from '@tensorflow/tfjs';

function initParams() {
  nst w1 = tf.randomUniform([10, 784], -0.5, 0.5);
  nst b1 = tf.randomUniform([10, 1], -0.5, 0.5);
  nst w2 = tf.randomUniform([10, 10], -0.5, 0.5);
  nst b2 = tf.randomUniform([10, 1], -0.5, 0.5);
  turn { w1, b1, w2, b2 };
}

function ReLU(z: tf.Tensor) {
  turn z.relu();
}

function softmax(z: tf.Tensor) {
  turn z.softmax();
}

function oneHot(y: number[], numClasses: number): tf.Tensor {
  turn tf.oneHot(tf.tensor1d(y, 'int32'), numClasses);
}

function forwardProp(w1: tf.Tensor, b1: tf.Tensor, w2: tf.Tensor, b2: tf.Tensor, x: tf.Tensor) {
  nst z1 = w1.matMul(x).add(b1);
  nst a1 = ReLU(z1);
  nst z2 = w2.matMul(a1).add(b2);
  nst a2 = softmax(z2);
  turn { z1, a1, z2, a2 };
}

function backProp(z1: tf.Tensor, a1: tf.Tensor, a2: tf.Tensor, w2: tf.Tensor, x: tf.Tensor, y: tf.Tensor) {
  nst m = y.shape[1];
  nst dz2 = a2.sub(y);
  nst dw2 = dz2.matMul(a1.transpose()).div(m);
  nst db2 = dz2.sum(1).reshape([10, 1]).div(m);
  nst dz1 = w2.transpose().matMul(dz2).mul(z1.greater(0));
  nst dw1 = dz1.matMul(x.transpose()).div(m);
  nst db1 = dz1.sum(1).reshape([10, 1]).div(m);
  turn { dw1, db1, dw2, db2 };
}

function updateParams(w1: tf.Tensor, b1: tf.Tensor, w2: tf.Tensor, b2: tf.Tensor, dw1: tf.Tensor, db1: tf.Tensor, dw2: tf.Tensor, db2: tf.Tensor, alpha: number) {
   = w1.sub(dw1.mul(alpha));
   = b1.sub(db1.mul(alpha));
   = w2.sub(dw2.mul(alpha));
   = b2.sub(db2.mul(alpha));
  turn { w1, b1, w2, b2 };
}

function getPredictions(a2: tf.Tensor) {
  turn a2.argMax(0);
}

function getAccuracy(predictions: tf.Tensor, y: tf.Tensor) {
  turn predictions.equal(y.argMax(0)).sum().div(y.shape[1]).arraySync();
}

async function gradientDescent(X: tf.Tensor, Y: tf.Tensor, iterations: number, alpha: number) {
  t { w1, b1, w2, b2 } = initParams();

  r (let i = 0; i < iterations; i++) {
    t { z1, a1, z2, a2 } = forwardProp(w1, b1, w2, b2, X);
    t { dw1, db1, dw2, db2 } = backProp(z1, a1, a2, w2, X, Y);
    1, b1, w2, b2 } = updateParams(w1, b1, w2, b2, dw1, db1, dw2, db2, alpha));

    i % 10 === 0) {
      predictions = getPredictions(a2);
      accuracy = getAccuracy(predictions, Y);
      e.log(`Iteration ${i}, Accuracy: ${accuracy}`);
    
  
  turn { w1, b1, w2, b2 };
}

function makePredictions(X: tf.Tensor, w1: tf.Tensor, b1: tf.Tensor, w2: tf.Tensor, b2: tf.Tensor) {
  nst { a2 } = forwardProp(w1, b1, w2, b2, X);
  turn getPredictions(a2);
}

async function main() {
  nst X_train = tf.randomUniform([784, 1000]);
  nst Y_train = oneHot([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 10).tile([1, 100]);

  nst { w1, b1, w2, b2 } = await gradientDescent(X_train, Y_train, 500, 0.1);
  nst X_test = tf.randomUniform([784, 10]);  // 10 test examples
  nst predictions = makePredictions(X_test, w1, b1, w2, b2);
  edictions.print();
}

main();

// Testing 

/*import * as tf from '@tensorflow/tfjs';

async function testNN() {
    const w1 = tf.randomUniform([10, 784], -0.5, 0.5);
    const b1 = tf.randomUniform([10, 1], -0.5, 0.5);
    const w2 = tf.randomUniform([10, 10], -0.5, 0.5);
    const b2 = tf.randomUniform([10, 1], -0.5, 0.5);

    const X_train = tf.randomUniform([784, 1000]);
    const Y_train = tf.oneHot(tf.tensor1d([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 'int32'), 10).tile([100, 1]);

    const { z1, a1, z2, a2 } = forwardProp(w1, b1, w2, b2, X_train);
    console.log("Forward Propagation successful");
}

testNN();
*/

File: CUDA/nn.py
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

data=pd.read_csv('/kaggle/input/digit-recognizer/train.csv')

data.head()

data=np.array(data)
m,n=data.shape
np.random.shuffle(data)

data_dev=data[0:1000].T
Y_dev=data_dev[0]
X_dev=data_dev[1:n]
X_dev=X_dev/255

data_train=data[1000:m].T
Y_train=data_train[0]
X_train=data_train[1:n]
X_train=X_train/255
_,m_train=X_train.shape

def init_params():
    w1=np.random.rand(10,784)-0.5
    b1=np.random.rand(10,1)-0.5
    w2=np.random.rand(10,10)-0.5
    b2=np.random.rand(10,1)-0.5
    return w1,b1,w2,b2

def ReLU(z):
    return np.maximum(0,z)

def softmax(Z):
    A = np.exp(Z) / sum(np.exp(Z))
    return A

def one_hot(z):
    one_hot_z=np.zeros((z.size,z.max()+1))
    one_hot_z[np.arange(z.size),z]=1
    one_hot_z=one_hot_z.T
    return one_hot_z

def derivative_ReLU(z):
    return z>0

def forward_prop(w1,b1,w2,b2,x):
    z1=w1.dot(x)+b1
    a1=ReLU(z1)
    z2=w2.dot(a1)+b2
    a2=softmax(z2)
    return z1,a1,z2,a2

def back_prop(z1,a1,z2,a2,w2,x,y):
    m=y.size
    one_hot_Y=one_hot(y)
    dz2=a2-one_hot_Y
    dw2=1/m * dz2.dot(a1.T)
    db2=1/m * np.sum(dz2)
    dz1 = w2.T.dot(dz2)*derivative_ReLU(z1)
    dw1=1/m * dz1.dot(x.T)
    db1=1/m * np.sum(dz1)
    return dw1,db1,dw2,db2

def update_params(w1,b1,w2,b2,dw1,db1,dw2,db2,alpha):
    w1=w1-alpha*dw1
    w2=w2-alpha*dw2
    b1=b1-alpha*db1
    b2=b2-alpha*db2
    return w1,b1,w2,b2

def get_predictions(a2):
    return np.argmax(a2,0)

def get_accuracy(predictions,y):
    print(predictions,y)
    return np.sum(predictions==y)/y.size

def gradient_descent(x,y,it,alpha):
    w1,b1,w2,b2=init_params()
    for i in range(it):
        z1,a1,z2,a2=forward_prop(w1,b1,w2,b2,x)
        dw1,db1,dw2,db2=back_prop(z1,a1,z2,a2,w2,x,y)
        w1,b1,w2,b2=update_params(w1,b1,w2,b2,dw1,db1,dw2,db2,alpha)
        if i%10==0:
            print("Iteration",i)
            print("Accuracy", get_accuracy(get_predictions(a2),y))
    return w1,b1,w2,b2

w1,b1,w2,b2=gradient_descent(X_train,Y_train,500,0.1)

def make_predictions(x,w1,b1,w2,b2):
    _,_,_,a2=forward_prop(w1,b1,w2,b2,x)
    predictions=get_predictions(a2)
    return predictions

def test_predictions(index,w1,b1,w2,b2):
    current_image=X_train[:,index,None]
    prediction=make_predictions(X_train[:,index,None],w1,b1,w2,b2)
    label=Y_train[index]
    print("Prediction : ",prediction)
    print("Label : ",label)
    current_image=current_image.reshape((28,28))*255
    plt.gray()
    plt.imshow(current_image,interpolation='nearest')
    plt.show()

dev_predictions=make_predictions(X_dev,w1,b1,w2,b2)
print("Accuracy - ",get_accuracy(dev_predictions,Y_dev))
